futures-channel/src/manual_reset_event.rs

```diff
@@ -0,0 +1,351 @@
+                }
+            },
+            PollState::Waiting => {
+                // The WaitHandle is already registered.
+                // The event can't have been set, since this would change the
+                // waitstate inside the mutex.
+                Poll::Pending
+            },
+            PollState::Done => {
+                // We had been woken up by the event.
+                // This does not guarantee that the event is still set. It could
+                // have been reset it in the meantime.
+                wait_handle.terminated = true;
+                Poll::Ready(())
+            },
+        }
+    }
+
+    fn remove_waiter(&mut self, wait_handle: &mut WaitHandleRegistration) {
+        let addr = wait_handle as *mut WaitHandleRegistration;
+
+        // WaitHandle only needs to get removed if it had been registered at
+        // the Event. This has happened in the PollState::Waiting case.
+        if let PollState::Waiting = wait_handle.state {
+            // Remove the WaitHandle from the linked list
+            if self.waiters == addr {
+                self.waiters = wait_handle.next;
+            } else {
+                // Find the WaitHandle before us and link it to the one
+                // behind us
+                let mut iter = self.waiters;
+                let mut found_addr = false;
+
+                unsafe {
+                    while !iter.is_null() {
+                        if (*iter).next == addr {
+                            (*iter).next = wait_handle.next;
+                            found_addr = true;
+                            break;
+                        } else {
+                            iter = (*iter).next;
+                        }
+                    }
+                }
+
+                // Panic if the address isn't found. This can only happen if the contract was
+                // violated, e.g. the WaitHandle got moved after the initial poll.
+                assert!(found_addr, "Future could not be unregistered");
+            }
+            wait_handle.next = null_mut();
+            wait_handle.state = PollState::Done;
+        }
+    }
+}
+
+/// Creates a new LocalManualResetEvent in the given state
+pub fn local_manual_reset_event(is_set: bool) -> LocalManualResetEvent {
+    LocalManualResetEvent {
+        inner: RefCell::new(InnerEventState::new(is_set)),
+    }
+}
+
+impl<'a> LocalManualResetEvent {
+    /// Sets the event.
+    ///
+    /// Setting the event will notify all pending waiters.
+    pub fn set(&self) {
+        self.inner.borrow_mut().set()
+    }
+
+    /// Resets the event.
+    pub fn reset(&self) {
+        self.inner.borrow_mut().reset()
+    }
+
+    /// Returns whether the event is set
+    pub fn is_set(&self) -> bool {
+        self.inner.borrow().is_set()
+    }
+
+    /// Returns a future that gets fulfilled when the event is set.
+    pub fn poll_set(&'a self) -> impl Future<Output = ()> + FusedFuture + 'a {
```

Comment:
Generally `poll_*` functions are ones that return `Poll<_>`, like `InnerEventWaiter::poll_waiter`. Is there some other appropriate name for this? Maybe just `wait`?

Reply:
That's right! I derived the name from the existing async code like AsyncRead. But since those are not really awaitable futures the situation is a bit different.
How about `get_awaiter()`? That's a bit aligned with the C# terminology. `get_future()` is obviously also ok, but less specific.